node-it/src/test/scala/com/wavesplatform/it/sync/transactions/TransferNFTSuite.scala

```diff
@@ -0,0 +1,167 @@
+         |func nftPaymentTransferToThirdAddress(address: String) = {
+         |    let thirdAddress = Address(fromBase58String(address))
+         |    let pmt = i.payment.extract()
+         |    TransferSet([
+         |            ScriptTransfer(thirdAddress, pmt.amount, pmt.assetId)
+         |        ])
+         |}
+         |
+         |@Callable(i)
+         |func transferAsPayment() = {
+         |    TransferSet([])
+         |    }
+         |
+         |@Callable(i)
+         |func nftTransferToSelf() = {
+         |    let pmt = i.payment.extract()
+         |    TransferSet([
+         |            ScriptTransfer(i.caller, pmt.amount, pmt.assetId)
+         |        ])
+         |}
+         |@Callable(i)
+         |func transferFromDappToAddress(address: String) = {
+         |    let recipient = Address(fromBase58String(address))
+         |    TransferSet([
+         |            ScriptTransfer(recipient, 1, base58'$nftAsset')
+         |        ])
+         |}
+         |@Verifier(t)
+         |func verify() = {
+         | true
+         |}
+        """.stripMargin
+    val script      = ScriptCompiler.compile(scriptText, ScriptEstimatorV2).explicitGet()._1.bytes().base64
+    sender.setScript(secondAddress, Some(script), setScriptFee, waitForTx = true)
+    def invokeTransfer(caller: String, functionName: String, args: List[Terms.EXPR] = List.empty, payment: Seq[InvokeScriptTransaction.Payment]): Transaction = {
+    sender.invokeScript(
+      caller,
+      secondAddress,
+      Some(functionName),
+      payment = payment,
+      args = args,
+      fee = 1300000,
+      waitForTx = true
+    )
+    }
+    for(functionName <- Seq(
+      "nftTransferToDapp",
+      "transferFromDappToAddress",
+      "nftTransferToSelf",
+      "nftPaymentTransferToThirdAddress",
+      "transferAsPayment")) {
+
+      val (caller, transferSender, recipient, functionArgs, payment) = functionName match {
+        case "nftTransferToDapp" => (firstAddress, firstAddress, secondAddress, List.empty, Seq(InvokeScriptTransaction.Payment(1, Asset.fromString(Some(nftAsset)))))
+        case "transferFromDappToAddress" => (firstAddress, secondAddress, thirdAddress, List(Terms.CONST_STRING(thirdAddress).explicitGet()), Seq.empty)
+        case "nftTransferToSelf" => (thirdAddress, secondAddress, thirdAddress, List.empty, Seq(InvokeScriptTransaction.Payment(1, Asset.fromString(Some(nftAsset)))))
+        case "nftPaymentTransferToThirdAddress" => (thirdAddress, thirdAddress, firstAddress, List(Terms.CONST_STRING(firstAddress).explicitGet()),
+          Seq(InvokeScriptTransaction.Payment(1, Asset.fromString(Some(nftAsset)))))
+        case "transferAsPayment" => (firstAddress, firstAddress, secondAddress, List.empty, Seq(InvokeScriptTransaction.Payment(1, Asset.fromString(Some(nftAsset)))))
+      }
+      invokeTransfer(caller, functionName, functionArgs, payment)
+      if(functionName.equals("nftPaymentTransferToThirdAddress")) {
```

Comment:
always check dapp's balance too

Reply:
It is actually checked for all cases here, but I rewrote the whole test so it became more readable.